### DISCUSSION TITLE
version 2.4.1 migration job "run-airflow-migrations" failed once, but it is not getting restarted

### DISCUSSION THREAD

**Discussion by subbugitdev** [apache/airflow#27992] [Q&A]
we are using apache airflow helmcharts with airflow version 2.4.1

The airflow-scheduler pod wait-for-airflow-migrations container was showing these errors and restarting every 60 seconds or so.

kubectl logs -f airflow-scheduler-6f6d976b8-p29t8 -c wait-for-airflow-migrations

/home/airflow/.local/lib/python3.7/site-packages/airflow/configuration.py:367: FutureWarning: The auth_backends setting in [api] has had airflow.api.auth.backend.session added in the running config, which is needed by the UI. Please update your config before Apache Airflow 3.0.
  FutureWarning,
[2022-11-29T15:42:30.517+0000] {db.py:737} INFO - Waiting for migrations... 0 second(s)
[2022-11-29T15:42:31.522+0000] {db.py:737} INFO - Waiting for migrations... 1 second(s)
..
[2022-11-29T16:08:36.566+0000] {db.py:737} INFO - Waiting for migrations... 59 second(s)
Traceback (most recent call last):
  File "/home/airflow/.local/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/__main__.py", line 39, in main
    args.func(args)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/cli/cli_parser.py", line 52, in command
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/cli/commands/db_command.py", line 138, in check_migrations
    db.check_migrations(timeout=args.migration_wait_timeout)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/db.py", line 739, in check_migrations
    f"There are still unapplied migrations after {timeout} seconds. Migration"
TimeoutError: There are still unapplied migrations after 60 seconds. MigrationHead(s) in DB: set() | Migration Head(s) in Source Code: {'ecb43d2a1842'}

The kubectl describe job airflow-run-airflow-migrations was not showing any detailed error message.
I had to manually restart the job by doing the following to stablizie things and get airflow scheduler, webserver start ok
kubectl get job airflow-run-airflow-migrations -o json | jq 'del(.spec.selector)' | jq 'del(.spec.template.metadata.labels)' | kubectl replace --force -f -
kubectl get job airflow-run-airflow-migrations -o yaml > airflow_run_migration.yaml
kubectl apply -f airflow_run_migration.yaml
Warning: resource jobs/airflow-run-airflow-migrations is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
job.batch/airflow-run-airflow-migrations configured

question is why the job did not get restarted onfailure ?

**Comment by subbugitdev** [comment]
airflow-run-airflow-migrations job was initally showing 1 failed run

**Reply by subbugitdev** [reply]
kubectl describe job airflow-run-airflow-migrations
Name:             airflow-run-airflow-migrations
Namespace:        default
Selector:         controller-uid=031f74be-2ac1-479d-8973-9cfa89d72d2c
Labels:           chart=airflow-1.7.0
                  component=run-airflow-migrations
                  heritage=Helm
                  kapp.k14s.io/app=1669705103632013816
                  kapp.k14s.io/association=v1.4545864af9583b17516077920faba34f
                  release=airflow
                  tier=airflow
Annotations:      helm.sh/hook: post-install,post-upgrade
                  helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
                  helm.sh/hook-weight: 1
                  kapp.k14s.io/identity: v1;default/batch/Job/airflow-run-airflow-migrations;batch/v1
                  kapp.k14s.io/original:
                    {"apiVersion":"batch/v1","kind":"Job","metadata":{"annotations":{"helm.sh/hook":"post-install,post-upgrade","helm.sh/hook-delete-policy":"...
                  kapp.k14s.io/original-diff-md5: 06cccbc0c144b0b6fb640cc6e60d5676
                  kbld.k14s.io/images:
                    - origins:
                      - resolved:
                          tag: 2.4.1
                          url: omega-docker-snapshot-local.artifactory.eng.vmware.com/omega-vsa/airflow:2.4.1
                      - preresolved:
                          url:xxxx
                      url: xxxx
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Start Time:       Tue, 29 Nov 2022 12:28:25 +0530
Pods Statuses:    0 Active / 0 Succeeded / 1 Failed
Pod Template:
  Labels:           component=run-airflow-migrations
                    controller-uid=031f74be-2ac1-479d-8973-9cfa89d72d2c
                    job-name=airflow-run-airflow-migrations
                    kapp.k14s.io/app=1669705103632013816
                    kapp.k14s.io/association=v1.4545864af9583b17516077920faba34f
                    release=airflow
                    tier=airflow
  Service Account:  airflow-migrate-database-job
  Containers:
   run-airflow-migrations:
    Image:      xxxx
    Port:       <none>
    Host Port:  <none>
    Args:
      bash
      -c
      exec \
      airflow db upgrade
    Environment:
      PYTHONUNBUFFERED:                               1
      AIRFLOW__CORE__LOAD_EXAMPLES:                   false
      AIRFLOW__WEBSERVER__WEB_SERVER_WORKER_TIMEOUT:  500
      AIRFLOW__WEBSERVER__WEB_SERVER_MASTER_TIMEOUT:  500
      AIRFLOW__CORE__DAGBAG_IMPORT_TIMEOUT:           500
      AIRFLOW__API__AUTH_BACKENDS:                    airflow.api.auth.backend.basic_auth
      AIRFLOW__CORE__FERNET_KEY:                      <set to the key 'fernet-key' in secret 'airflow-fernet-key'>                      Optional: false
      AIRFLOW__CORE__SQL_ALCHEMY_CONN:                <set to the key 'connection' in secret 'airflow-service-secrets'>                 Optional: false
      AIRFLOW__DATABASE__SQL_ALCHEMY_CONN:            <set to the key 'connection' in secret 'airflow-service-secrets'>                 Optional: false
      AIRFLOW_CONN_AIRFLOW_DB:                        <set to the key 'connection' in secret 'airflow-service-secrets'>                 Optional: false
      AIRFLOW__WEBSERVER__SECRET_KEY:                 <set to the key 'webserver-secret-key' in secret 'airflow-webserver-secret-key'>  Optional: false
    Mounts:
      /opt/airflow/airflow.cfg from config (ro,path="airflow.cfg")
  Volumes:
   config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      airflow-airflow-config
    Optional:  false
Events:        <none>

**Reply by subbugitdev** [reply]
generated helm-chart shows the job having restartpolicy of onFailure ..

`################################
## Airflow Run Migrations
#################################
apiVersion: batch/v1
kind: Job
metadata:
  name: release-name-run-airflow-migrations
  labels:
    tier: airflow
    component: run-airflow-migrations
    release: release-name
    chart: "airflow-1.7.0"
    heritage: Helm
  annotations:
    helm.sh/hook: post-install,post-upgrade
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
    helm.sh/hook-weight: "1"
spec:
  template:
    metadata:
      labels:
        tier: airflow
        component: run-airflow-migrations
        release: release-name
    spec:
      securityContext: 
        runAsUser: 50000
        fsGroup: 0
      restartPolicy: OnFailure`

**Reply by potiuk** [reply]
Because this is the default timeout:

https://airflow.apache.org/docs/helm-chart/stable/parameters-ref.html#database

<img width="1281" alt="Screenshot 2022-11-29 at 20 14 10" src="https://user-images.githubusercontent.com/595491/204624285-ace538ed-2b98-4b38-b9cb-03da38d6a0ff.png">


If your migration needs to take more time you need to give it more time. If your migration fails, you need to find out (in the logs of that job) why it failed). This should give you more information to allow you to determine the root cause.

Also make sure that you have for sure the right airflow version in your images. Sometimes (especially when people try to attempt to install packages dynamically when their images are starting - which you should  never do in production), they might acccidentally downgrade or upgrade airflow and then the migration might fail or behave strangely. So make sure your images that you are using are "static" (i.e. no packages updated/installed after they start) and that they have the right airflow version.

**Reply by subbugitdev** [reply]
Thanks for the response.
My concern was why the failed job didn't get restarted, despite the setting for restartPolicy: OnFailure .
If the cluster does not detect and restart the job, no timeout will be large enough.
We will try to see if we can reproduce this issue again to see if was a one-off/random occurence.

**Reply by potiuk** [reply]
Migration should never be restarted. I don't remember how it is defined but migration Jo is not supposed to be continuously run or restarted. It should be run once at deployment time and either let the deployment continue when succeeded or fail. Restarting failed migration job makes rather little sense - most likely it will fail again

**Comment by minnieshi** [comment]
Did you use flux/argocd/terraform? Or did you use kubectl to install/upgrade?

On this page at the bottom it says 
one needs to disable some hook if using flux/argocd/terraform 
https://airflow.apache.org/docs/helm-chart/stable/index.html

> we are using apache airflow helmcharts with airflow version 2.4.1
> 
> 
> The airflow-scheduler pod wait-for-airflow-migrations container was showing these errors and restarting every 60 seconds or so.
> 
> 
> 
> kubectl logs -f airflow-scheduler-6f6d976b8-p29t8 -c wait-for-airflow-migrations
> 
> 
> /home/airflow/.local/lib/python3.7/site-packages/airflow/configuration.py:367: FutureWarning: The auth_backends setting in [api] has had airflow.api.auth.backend.session added in the running config, which is needed by the UI. Please update your config before Apache Airflow 3.0.
> 
>   FutureWarning,
> 
> [2022-11-29T15:42:30.517+0000] {db.py:737} INFO - Waiting for migrations... 0 second(s)
> 
> [2022-11-29T15:42:31.522+0000] {db.py:737} INFO - Waiting for migrations... 1 second(s)
> 
> ..
> 
> [2022-11-29T16:08:36.566+0000] {db.py:737} INFO - Waiting for migrations... 59 second(
>

...
> The kubectl describe job airflow-run-airflow-migrations was not showing any detailed error message.
> 
> I had to manually restart the job by doing the following to stablizie things and get airflow scheduler, webserver start ok
> 
> kubectl get job airflow-run-airflow-migrations -o json | jq 'del(.spec.selector)' | jq 'del(.spec.template.metadata.labels)' | kubectl replace --force -f -
> 
> kubectl get job airflow-run-airflow-migrations -o yaml > airflow_run_migration.yaml
> 
> kubectl apply -f airflow_run_migration.yaml

**Reply by minnieshi** [reply]
I found my migration job would not restart for the 2nd time (the 1st time run was when the default airflow is deployed onto Kubernetes and it had no issues),  and then i started to apply changes to the values.yaml file such as **make the database to be azure postgresql**; but then it would not take the values into effect, see screen shots; 

of course my debug skills on kubernetes are not high, so i would need extra help if extra info is needed.

![image](https://user-images.githubusercontent.com/11322886/209687297-7d83e4aa-9096-467e-851a-2557928da2b6.png)
![image](https://user-images.githubusercontent.com/11322886/209687323-fc853fcc-438c-4bea-8182-793dac722cae.png)
![image](https://user-images.githubusercontent.com/11322886/209687349-5c043188-3393-49b2-a73f-a997e55d6c3c.png)

```
      database:
        sql_alchemy_conn_secret: airflow-postgres-redis
        sql_alchemy_connect_args:
          {
            "keepalives": 1,
            "keepalives_idle": 30,
            "keepalives_interval": 5,
            "keepalives_count": 5,
          }
    postgresql:
      enabled: false
    pgbouncer:
      enabled: false
    # Airflow database & redis config
    data:
      metadataSecretName: airflow-postgres-redis
```
check again the pod for waiting for the migration: 
![image](https://user-images.githubusercontent.com/11322886/209689640-fdeed08d-19b3-43d5-a736-466cf36237ba.png)

and below was the 1st success at the initial installation (which did not use external db)
```
kubectl describe job airflow-airflow-run-airflow-migrations
Name:             airflow-airflow-run-airflow-migrations
Namespace:        airflow
Selector:         controller-uid=efdc3c7b-5172-4841-abcf-17e055fa6e2e
Labels:           app.kubernetes.io/managed-by=Helm
                  chart=airflow-1.7.0
                  component=run-airflow-migrations
                  helm.toolkit.fluxcd.io/name=airflow
                  helm.toolkit.fluxcd.io/namespace=airflow
                  heritage=Helm
                  release=airflow-airflow
                  tier=airflow
Annotations:      batch.kubernetes.io/job-tracking:
                  meta.helm.sh/release-name: airflow-airflow
                  meta.helm.sh/release-namespace: airflow
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Start Time:       Tue, 27 Dec 2022 14:21:50 +0100
Completed At:     Tue, 27 Dec 2022 14:22:29 +0100
Duration:         39s
Pods Statuses:    0 Active (0 Ready) / 1 Succeeded / 0 Failed
Pod Template:
  Labels:           component=run-airflow-migrations
                    controller-uid=efdc3c7b-5172-4841-abcf-17e055fa6e2e
                    job-name=airflow-airflow-run-airflow-migrations
                    release=airflow-airflow
                    tier=airflow
  Service Account:  airflow-airflow-migrate-database-job
  Containers:
   run-airflow-migrations:
    Image:      apache/airflow:2.4.1
    Port:       <none>
    Host Port:  <none>
    Args:
      bash
      -c
      exec \
      airflow db upgrade
    Environment:
      PYTHONUNBUFFERED:                     1
      AIRFLOW__CORE__FERNET_KEY:            <set to the key 'fernet-key' in secret 'airflow-airflow-fernet-key'>                      Optional: false
      AIRFLOW__CORE__SQL_ALCHEMY_CONN:      <set to the key 'connection' in secret 'airflow-airflow-airflow-metadata'>                Optional: false
      AIRFLOW__DATABASE__SQL_ALCHEMY_CONN:  <set to the key 'connection' in secret 'airflow-airflow-airflow-metadata'>                Optional: false
      AIRFLOW_CONN_AIRFLOW_DB:              <set to the key 'connection' in secret 'airflow-airflow-airflow-metadata'>                Optional: false
      AIRFLOW__WEBSERVER__SECRET_KEY:       <set to the key 'webserver-secret-key' in secret 'airflow-airflow-webserver-secret-key'>  Optional: false
      AIRFLOW__CELERY__BROKER_URL:          <set to the key 'connection' in secret 'airflow-airflow-broker-url'>                      Optional: false
    Mounts:
      /opt/airflow/airflow.cfg from config (ro,path="airflow.cfg")
  Volumes:
   config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      airflow-airflow-airflow-config
    Optional:  false
Events:        <none>
```

**Reply by minnieshi** [reply]
> I found my migration job would not restart for the 2nd time (the 1st time run was when the default airflow is deployed onto Kubernetes and it had no issues), and then i started to apply changes to the values.yaml file such as **make the database to be azure postgresql**; but then it would not take the values into effect, see screen shots;
> 

it turns out my helmRelease kind in the flux spec 'timeout' was set too short for the migration to even finish i think.

**Reply by minnieshi** [reply]
> lmRelease kind in the flux spec 'timeout' was set too short

my further experiment/try tells me it is NOT yet conclusive 'timeout' too short was the cause;
it shows the 2 jobs were still not started successfully. it was only run once.
more trust worthy test will need to be done with a bit help, such as what kind of changes will trigger migration job to run. as i found my above false conclusion was the deployment was only twice.
see below helm release history; and i could not make the 3rd release to succeed even though the values are 100% correct; so **the bug/issue short description is helmRelease in combination with `flux` have issues with db migration jobs (only run once <can be successful>) which makes it a stopper for further upgrade**

```
REVISION	UPDATED                 	STATUS    	CHART        	APP VERSION	DESCRIPTION
1       	Wed Dec 28 02:22:42 2022	superseded	airflow-1.7.0	2.4.1      	Install complete
2       	Wed Dec 28 02:43:25 2022	deployed  	airflow-1.7.0	2.4.1      	Upgrade complete
```


see below equivalent values , even tried to disable the db migration did not make flux to work with it.
```
createUserJob:
  useHelmHooks: false
migrateDatabaseJob:
  useHelmHooks: false
config:
  webserver:
    expose_config: 'non-sensitive-only'
postgresql:
  enabled: false
pgbouncer:
  enabled: true
  # The maximum number of connections to PgBouncer
  maxClientConn: 100
  # The maximum number of server connections to the metadata database from PgBouncer
  metadataPoolSize: 10
  # The maximum number of server connections to the result backend database from PgBouncer
  resultBackendPoolSize: 5
# Airflow database & redis config
data:
  metadataSecretName: airflow-postgres-redis
# to generate strong secret: python3 -c 'import secrets; print(secrets.token_hex(16))'
webserverSecretKeySecretName: airflow-webserver-secret

```

and see below 2 jobs
```
$ kubectl describe job -n airflow
Name:             airflow-airflow-create-user
Namespace:        airflow
Selector:         controller-uid=8b09e28b-ba3a-4cee-b20f-693a3aa15363
Labels:           app.kubernetes.io/managed-by=Helm
                  chart=airflow-1.7.0
                  component=create-user-job
                  helm.toolkit.fluxcd.io/name=airflow
                  helm.toolkit.fluxcd.io/namespace=airflow
                  heritage=Helm
                  release=airflow-airflow
                  tier=airflow
Annotations:      batch.kubernetes.io/job-tracking:
                  meta.helm.sh/release-name: airflow-airflow
                  meta.helm.sh/release-namespace: airflow
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Start Time:       Wed, 28 Dec 2022 03:22:46 +0100
Completed At:     Wed, 28 Dec 2022 03:24:32 +0100
Duration:         106s
Pods Statuses:    0 Active (0 Ready) / 1 Succeeded / 0 Failed
Pod Template:
  Labels:           component=create-user-job
                    controller-uid=8b09e28b-ba3a-4cee-b20f-693a3aa15363
                    job-name=airflow-airflow-create-user
                    release=airflow-airflow
                    tier=airflow
  Service Account:  airflow-airflow-create-user-job
  Containers:
   create-user:
    Image:      apache/airflow:2.4.1
    Port:       <none>
    Host Port:  <none>
    Args:
      bash
      -c
      exec \
      airflow users create "$@"
      --
      -r
      Admin
      -u
      admin
      -e
      admin@example.com
      -f
      admin
      -l
      user
      -p
      admin
    Environment:
      AIRFLOW__CORE__FERNET_KEY:            <set to the key 'fernet-key' in secret 'airflow-airflow-fernet-key'>                      Optional: false
      AIRFLOW__CORE__SQL_ALCHEMY_CONN:      <set to the key 'connection' in secret 'airflow-postgres-redis'>                          Optional: false
      AIRFLOW__DATABASE__SQL_ALCHEMY_CONN:  <set to the key 'connection' in secret 'airflow-postgres-redis'>                          Optional: false
      AIRFLOW_CONN_AIRFLOW_DB:              <set to the key 'connection' in secret 'airflow-postgres-redis'>                          Optional: false
      AIRFLOW__WEBSERVER__SECRET_KEY:       <set to the key 'webserver-secret-key' in secret 'airflow-airflow-webserver-secret-key'>  Optional: false
      AIRFLOW__CELERY__BROKER_URL:          <set to the key 'connection' in secret 'airflow-airflow-broker-url'>                      Optional: false
    Mounts:
      /opt/airflow/airflow.cfg from config (ro,path="airflow.cfg")
  Volumes:
   config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      airflow-airflow-airflow-config
    Optional:  false
Events:        <none>


Name:             airflow-airflow-run-airflow-migrations
Namespace:        airflow
Selector:         controller-uid=5da8c81f-7920-4eaf-9d7a-58a48c740bdc
Labels:           app.kubernetes.io/managed-by=Helm
                  chart=airflow-1.7.0
                  component=run-airflow-migrations
                  helm.toolkit.fluxcd.io/name=airflow
                  helm.toolkit.fluxcd.io/namespace=airflow
                  heritage=Helm
                  release=airflow-airflow
                  tier=airflow
Annotations:      batch.kubernetes.io/job-tracking:
                  meta.helm.sh/release-name: airflow-airflow
                  meta.helm.sh/release-namespace: airflow
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Start Time:       Wed, 28 Dec 2022 03:22:46 +0100
Completed At:     Wed, 28 Dec 2022 03:23:07 +0100
Duration:         21s
Pods Statuses:    0 Active (0 Ready) / 1 Succeeded / 0 Failed
Pod Template:
  Labels:           component=run-airflow-migrations
                    controller-uid=5da8c81f-7920-4eaf-9d7a-58a48c740bdc
                    job-name=airflow-airflow-run-airflow-migrations
                    release=airflow-airflow
                    tier=airflow
  Service Account:  airflow-airflow-migrate-database-job
  Containers:
   run-airflow-migrations:
    Image:      apache/airflow:2.4.1
    Port:       <none>
    Host Port:  <none>
    Args:
      bash
      -c
      exec \
      airflow db upgrade
    Environment:
      PYTHONUNBUFFERED:                     1
      AIRFLOW__CORE__FERNET_KEY:            <set to the key 'fernet-key' in secret 'airflow-airflow-fernet-key'>                      Optional: false
      AIRFLOW__CORE__SQL_ALCHEMY_CONN:      <set to the key 'connection' in secret 'airflow-postgres-redis'>                          Optional: false
      AIRFLOW__DATABASE__SQL_ALCHEMY_CONN:  <set to the key 'connection' in secret 'airflow-postgres-redis'>                          Optional: false
      AIRFLOW_CONN_AIRFLOW_DB:              <set to the key 'connection' in secret 'airflow-postgres-redis'>                          Optional: false
      AIRFLOW__WEBSERVER__SECRET_KEY:       <set to the key 'webserver-secret-key' in secret 'airflow-airflow-webserver-secret-key'>  Optional: false
      AIRFLOW__CELERY__BROKER_URL:          <set to the key 'connection' in secret 'airflow-airflow-broker-url'>                      Optional: false
    Mounts:
      /opt/airflow/airflow.cfg from config (ro,path="airflow.cfg")
  Volumes:
   config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      airflow-airflow-airflow-config
    Optional:  false
Events:        <none>
```

**Reply by minnieshi** [reply]
I've now created a bug: https://github.com/apache/airflow/issues/28637

**Comment by subbugitdev** [comment]
@minnieshi 
For my case, i was able to restart the job and it ran ok.
1st save yaml
kubectl get job airflow-run-airflow-migrations -o yaml > airflow_run_migration.yaml

2nd tried this ...
kubectl get job airflow-run-airflow-migrations -o json | jq 'del(.spec.selector)' | jq 'del(.spec.template.metadata.labels)' | kubectl replace --force -f -

then this ..
# kubectl apply -f airflow_run_migration.yaml

checked it ran OK with
# kubectl get job -o wide |grep migration

**Comment by qx-vivek-bhosale** [comment]
I have deployed the below job 
```yaml
# Airflow Run Migrations
apiVersion: batch/v1
kind: Job
metadata:
  name: dev-3pp-airflow-run-airflow-migrations
  namespace: dev-3pp-airflow
  labels:
    tier: airflow
    component: run-airflow-migrations
    release: dev-3pp-airflow
    chart: "airflow-1.13.1"
  heritage: Helm
  annotations:
    helm.sh/hook: post-install,post-upgrade
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
    helm.sh/hook-weight: "1"
spec:
  template:
    metadata:
      labels:
        tier: airflow
        component: run-airflow-migrations
        release: dev-3pp-airflow
    spec:
      securityContext:
        runAsUser: 50000
        fsGroup: 0
      restartPolicy: OnFailure

```

and also in my PostgreSQL pod getting below error -

Traceback (most recent call last):
  File "/home/airflow/.local/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/__main__.py", line 57, in main
    args.func(args)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/cli_config.py", line 49, in command
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/providers_configuration_loader.py", line 55, in wrapped_function
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/commands/db_command.py", line 190, in check_migrations
    db.check_migrations(timeout=args.migration_wait_timeout)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/db.py", line 800, in check_migrations
    raise TimeoutError(
TimeoutError: There are still unapplied migrations after 60 seconds. MigrationHead(s) in DB: set() | Migration Head(s) in Source Code: {'88344c1d9134'}

It seems I am having below issue due to 

There are still unapplied migrations after 60 seconds.
Migration Head(s) in Source Code: {'88344c1d9134'} (This means there's a migration with ID '88344c1d9134' present in your Airflow code but not applied to the database)

can you help me out here ?

**Reply by subbugitdev** [reply]
In my case, only saw this during chaos testing (i.e. QA bringing down random k8s worker nodes) i.e. not during normal startup of Airflow. I suggest you can try the manual workaround that I suggested earlier (link below)
https://github.com/apache/airflow/discussions/27992#discussioncomment-4584447
